File: 백준/Gold/1717. 집합의 표현/집합의 표현.py
def find(a) :
    if a != parent[a] :
        parent[a] = find(parent[a])
    return parent[a]

def union(a, b) :
    a = find(a)
    b = find(b)
    if (a != b) :
        parent[b] = a

n, m = map(int, input().split())
parent = [i for i in range(n + 1)]

for _ in range(m) :
    cmd, a, b = map(int, input().split())
    if cmd == 0 :
        union(a, b)
    elif cmd == 1 :
        if find(a) == find(b) :
            print("YES")
        else :
            print("NO")

File: 백준/Gold/1717. 집합의 표현/집합의 표현.java
import java.io.*;
import java.util.*;

public class Main {
	static int parent[];
	public static void main(String[] args) throws IOException {
		Scanner sc = new Scanner(System.in);
		int N = sc.nextInt();
		int M = sc.nextInt();
		parent = new int[N + 1];
		for(int i = 0; i <= N; i++) {
			parent[i] = i;
		}
		
		for(int i = 0; i < M; i++) {
			int question = sc.nextInt();
			int a = sc.nextInt();
			int b = sc.nextInt();
			if (question == 0) {
				union(a, b);
			} else {
				if(find(a) == find(b)) {
					System.out.println("YES");
				} else {
					System.out.println("NO");
				}
			}
		}
	}

	private static int find(int a) {
		if (a == parent[a]) {
			return a;
		}
		
		return parent[a] = find(parent[a]);
	}

	private static void union(int a, int b) {
		a = find(a);
		b = find(b);
		
		if (a != b) {
			parent[b] = a;
		}
	}
}